workflow-ui/core-android/src/main/java/com/squareup/workflow1/ui/androidx/StateRegistryAggregator.kt

```diff
@@ -0,0 +1,300 @@
+   */
+  public fun attachToParentRegistry(
+    key: String,
+    parentOwner: SavedStateRegistryOwner
+  ) {
+    // attachToParentRegistry may be called multiple times without a matching detach in some cases,
+    // eg. when certain UI tests have failed and are being torn down. Ensure that if that happens
+    // we detach from the previous parent first.
+    detachFromParentRegistry()
+
+    this.parentRegistryOwner = parentOwner
+    this.parentKey = key
+
+    // We can only be restored once.
+    if (isRestored) return
+
+    val parentRegistry = parentOwner.savedStateRegistry
+    val parentLifecycle = parentOwner.lifecycle
+
+    // If the key is already registered, SavedStateRegistry throws an exception that doesn't provide
+    // enough information to troubleshoot, so we add some ourselves.
+    try {
+      parentRegistry.registerSavedStateProvider(key, ::saveToBundle)
+    } catch (e: IllegalArgumentException) {
+      throw IllegalArgumentException(
+        "Error registering SavedStateProvider: key \"$key\" is already in " +
+          "use on parent SavedStateRegistryOwner $parentOwner.\n" +
+          "This is most easily remedied by giving your container Screen rendering a unique " +
+          "Compatible.compatibilityKey, perhaps by wrapping it with Named.",
+        e
+      )
+    }
+
+    // Even if the parent lifecycle is in a state further than CREATED, new observers are sent all
+    // the lifecycle events required to catch the observer up to the current state, so this handles
+    // both the cases where we're ready to immediately restore, and where we have to wait.
+    parentLifecycle.addObserver(lifecycleObserver)
+  }
+
+  /**
+   * Must be called when the owning view detaches from the window.
+   *
+   * Stops listening to the parent lifecycle and unregisters from the parent registry.
+   */
+  public fun detachFromParentRegistry() {
+    // parentKey will only be null if parentRegistryOwner is also null.
+    parentRegistryOwner?.savedStateRegistry?.unregisterSavedStateProvider(parentKey!!)
+    parentRegistryOwner?.lifecycle?.removeObserver(lifecycleObserver)
+    parentRegistryOwner = null
+    parentKey = null
+    children.clear()
+  }
+
+  /**
+   * Puts a new [ViewTreeSavedStateRegistryOwner] in place on [view], registered
+   * with its [WorkflowLifecycleOwner]. **This method must be called before [view]
+   * is attached to a window.**
+   *
+   * Clean up requirements after making this call are nuanced:
+   *
+   *  - all instances are dropped after [detachFromParentRegistry] is called
+   *
+   *  - if [view] is dropped but may be restored before [detachFromParentRegistry] is called,
+   *    as when pushing and popping a back stack, call [saveChildRegistryOwner] to capture
+   *    its state tear down its [ViewTreeSavedStateRegistryOwner]. The saved state
+   *    will be restored to the next [view] passed to [installChildSavedStateRegistryOwnerOn]
+   *    with the same [key]
```

Comment:
Got lost in the grammar of the first sentence?  "call X to capture its state tear down its ..."?

Reply:
Is this better?

```
   *  - if [view] is dropped but may later be replaced with a new instance, as when
   *   pushing and popping a back stack, call [saveAndPruneChildRegistryOwner]
   *   to capture its state, and to stop requesting updates from it.
   *   The saved state will be restored to the next [view] passed to
   *   [installChildRegistryOwnerOn] with the same [key]
```